### PR TITLE
i18n support

### PR DESCRIPTION
**Pull request by pukmajster**
### Example usage

The editor can be localised by passing the `i18n` prop to the editor like so:

```tsx
const slovenianLocale: MDXEditorI18n = {
  toolbar: {
    blockTypeSelect: {
      selectBlockTypeTooltip: 'Izberi vrsto bloka',
      placeholder: 'Vrsta bloka'
    },

    blockTypes: {
      paragraph: 'Odstavek',
      heading: 'Naslov',
      quote: 'Citat'
    }
  }
}

function LocalizedEditor() {
  return <MDXEditor i18n={slovenianLocale} />
}
```

This code above will change the strings for the block type select element and its labels:

image

### How it's done internally

i18n is handled via a React context. More specifically, the `useI18n` hook. A plugin can access this hook and it will have all the necessary translations available to it

```tsx
export default function ExampleElement() {
  const i18n = useI18n()
  return <button title={i18n.toolbar.bold}>{i18n.toolbar.bold}</button>
}
```

--- 

I may have missed a translation or two, and some additional work may be required for custom plugins, but I think this is a decent starting point. I will happily provide additional development on the subject, so please, let me know what you think.

### CODE REVIEW

**Comment by petyosi** [comment]
Thank you very much for that contribution. This is indeed a great addition. I pushed a change that does the following:

- Got rid of the provider, moving the i18n object into the built-in state management flow. One context less to deal with.
- the `@types` dir is meant for overriding types in external libraries. I squashed all into core.

With the changes above, any plugin should be able to access the hook or the `i18n$` cell (if they need it outside of React). 

To extend further on the topic of external plugins needing more translations: I think it makes sense to take advantage of TypeScript's capability [to merge interface declarations](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces). 

For that purpose, I think we should use interfaces rather than types (a small example in the docs on how to declare the interface from the outside would be great - here's one example where I do that for an external interface: https://github.com/mdx-editor/editor/blob/main/src/mdastUtilHtmlComment.ts#L9-L15).  Let me know if this makes sense.

**Comment by pukmajster** [comment]
Yeah agreed, interfaces are better for this. I've gone ahead and converted the types. Is this then, in theory, enough for external plugins to fully utilise i18n?

**Comment by petyosi** [comment]
That looks good to me. I will give it another thorough read and merge next week.

File: src/plugins/core/i18n.ts [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+ */
+type DeepPartial<T> = T extends object
+  ? {
+      [P in keyof T]?: DeepPartial<T[P]>
+    }
+  : T
+
+export interface MDXEditorI18nPartial extends DeepPartial<MDXEditorI18n> {}
+
+export const defaultEnglishI18n: MDXEditorI18n = {
```
Comment (StanlieK) [inline src/plugins/core/i18n.ts:150]: I would expect that the default translations tell what to MDXEditorI18n type should look like. In that case, you don't have to care whether you extend the type or not.

```
type MDXEditor18n = typeof defaultEnglishI18n;
```

**Review by StanlieK — APPROVED** [review]

**Comment by petyosi** [comment]
I've given this some more thought. I will work on distributing the translation pieces into each plugin to see how this could work. In that way, each plugin should be able to include its own parts conveniently. Of course, the final translation override should happen at the root component prop.

**Pull request closed without merging**